Add shared key auth to data_lake

thovoll opened the pull request.

cataggar reviewed (commented):
Is this just a rename that matches other SDKs?

thovoll commented:
> Is this just a rename that matches other SDKs?

No, small prefactoring for adding shared key auth, which we need downstream in delta-rs.

heaths commented:
> > Is this just a rename that matches other SDKs?
> 
> No, small prefactoring for adding shared key auth, which we need downstream in delta-rs.

This be called StoragedSharedKeyCredential and follow a [similar pattern](https://docs.microsoft.com/dotnet/api/azure.storage.storagesharedkeycredential.-ctor?view=azure-dotnet#Azure_Storage_StorageSharedKeyCredential__ctor_System_String_System_String_) as with our other SDKs.

If we yet have a common storage crate, I'd put it in there as well.

heaths commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,230 @@
+use std::sync::Arc;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SharedKeyAuthorizationPolicy {
```
Should be `StorageSharedKeyCredential` that is public, which takes `account_name` then `account_key`. The policy itself shouldn't need to be public (at least, not exported from the crate but perhaps just from this module). The associated builder should then take the `StorageSharedKeyCredential` as an option for authenticating.

thovoll replied:
Thanks, this is definitely the longer term goal. Short term, I'm just switching from Bearer token to Shared key in the quickest way possible to unblock another project.

thovoll replied:
@heaths please take another look now. I introduced the StorageSharedKeyCredential struct.

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+
+        let uri_path = &request.uri().path_and_query().unwrap().to_string()[1..];
+        let full_url = format!("{}/{}", self.base_url, uri_path);
+        // println!(
```
This should be removed right?

thovoll replied:
Yes, these should be replaced with unit tests. For this PR, I'd like to change as little as possible in this code, since it's directly copied from the pre-pipeline shared key auth code in storage_account_client.rs. If there are any issues there, I'd like to address them in a follow-up PR.

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+        headers_mut.append(
+            azure_core::headers::MS_DATE,
+            HeaderValue::from_str(
+                format!("{}", chrono::Utc::now().format("%a, %d %h %Y %T GMT")).as_str(),
```
Does this work?
```suggestion
                &chrono::Utc::now().format("%a, %d %h %Y %T GMT").to_string()
```

thovoll replied:
For this PR, I'd like to change as little as possible in this code, since it's directly copied from the pre-pipeline shared key auth code in storage_account_client.rs. If there are any issues there, I'd like to address them in a follow-up PR.

thovoll replied:
Yes, fixed in follow-up PR.

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+            HeaderValue::from_str("2019-12-12")?,
+        ); // TODO: Remove duplication with storage_account_client.rs
+
+        let uri_path = &request.uri().path_and_query().unwrap().to_string()[1..];
```
Do we want to panic when path_and_query is `None`?

thovoll replied:
Probably not :) Code lifted from the Cosmos auth policy, so same problem there: https://github.com/Azure/azure-sdk-for-rust/blob/ad4805942985772e2d3f1adbc08a45eb16d7a48a/sdk/cosmos/src/authorization_policy.rs#L57

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+) -> String {
+    let str_to_sign = string_to_sign(http_headers, url, http_method, storage_account_name);
+
+    // println!("\nstr_to_sign == {:?}\n", str_to_sign);
+    // debug!("str_to_sign == {}", str_to_sign);
```
Can you remove all of these?

thovoll replied:
Yes, want to replace with unit tests

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+        .map(|s| if s == "0" { "" } else { s.to_str().unwrap() })
+        .unwrap_or("");
+    format!(
+        "{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}{}",
```
Are the extra newlines allowed if a header doesn't exist?

thovoll replied:
They are required actually, see here: https://docs.microsoft.com/en-us/rest/api/storageservices/authorize-with-shared-key#blob-queue-and-file-services-shared-key-authorization

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+
+fn add_if_exists<K: http::header::AsHeaderName>(h: &HeaderMap, key: K) -> &str {
+    match h.get(key) {
+        Some(ce) => ce.to_str().unwrap(),
```
I assume we should handle the case if the header is not UTF-8

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+fn canonicalize_header(h: &HeaderMap) -> String {
+    let mut v_headers = h
+        .iter()
+        .filter(|(k, _v)| k.as_str().starts_with("x-ms"))
+        .map(|(k, _)| k.as_str())
```
You can combine these with `filter_map`

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+
+fn canonicalized_resource(account: &str, u: &url::Url) -> String {
+    let mut can_res: String = String::new();
+    can_res += "/";
+    can_res += account;
```
Same as above

thovoll replied:
For this PR, I'd like to change as little as possible in this code, since it's directly copied from the pre-pipeline shared key auth code in storage_account_client.rs. If there are any issues there, I'd like to address them in a follow-up PR.

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+}
+
+fn canonicalized_resource(account: &str, u: &url::Url) -> String {
+    let mut can_res: String = String::new();
```
```suggestion
    let mut can_res = String::new();
```

thovoll replied:
For this PR, I'd like to change as little as possible in this code, since it's directly copied from the pre-pipeline shared key auth code in storage_account_client.rs. If there are any issues there, I'd like to address them in a follow-up PR.

thovoll replied:
Done

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+                trace!("adding to qps {:?}", q);
+
+                // add only once
+                if !(qps.iter().any(|x: &String| x == q.as_ref())) {
```
Use a `HashSet` instead of a `Vec` and you uniqueness is guaranteed.

thovoll replied:
For this PR, I'd like to change as little as possible in this code, since it's directly copied from the pre-pipeline shared key auth code in storage_account_client.rs. If there are any issues there, I'd like to address them in a follow-up PR.

rylev commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+) -> Vec<std::borrow::Cow<'a, str>> {
+    let mut v_values = Vec::new();
+
+    for item in vec.filter(|x| x.0 == *query_param) {
+        v_values.push(item.1)
+    }
```
```suggestion
    let v_values = vec.filter(|x| x.0 == *query_param).map(|item| item.1).collect() ;
```

thovoll replied:
For this PR, I'd like to change as little as possible in this code, since it's directly copied from the pre-pipeline shared key auth code in storage_account_client.rs. If there are any issues there, I'd like to address them in a follow-up PR.

rylev reviewed (changes requested):
Not sure how done this is, and I can't comment on the correctness, but I had a few comments on code quality.

thovoll commented:
> Not sure how done this is, and I can't comment on the correctness, but I had a few comments on code quality.

Thanks @rylev! Will review. The current state is that the e2e test passes (haven't tried the examples but should pass as well). I'm looking into doing a bit more work to align with the .Net SDK as proposed by @heaths. Overall goal is to quickly enable delta-rs PR to move forward and follow-up here with more small PRs to move the auth forward.

thovoll commented on sdk/storage/src/core/clients/storage_account_client.rs:
```diff
@@ -26,8 +26,10 @@ pub const EMULATOR_ACCOUNT: &str = "devstoreaccount1";
     "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==";
 
+// TODO: Is this used outside of here?
 pub(crate) const HEADER_VERSION: &str = "x-ms-version";
```
Fixed.

thovoll commented on sdk/storage_datalake/src/clients/data_lake_client.rs:
```diff
@@ -41,7 +42,9 @@ impl DataLakeClient {
         let per_call_policies = Vec::new();
         let auth_policy: Arc<dyn azure_core::Policy> =
-            Arc::new(AuthorizationPolicy::new(bearer_token));
+            // TODO: Make parameterized
```
Clarified.

thovoll commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+}
+
+impl SharedKeyAuthorizationPolicy {
+    #[allow(unused)]
```
Fixed.

thovoll commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+
+    for header_name in v_headers {
+        let s = h.get(header_name).unwrap().to_str().unwrap();
+        can = can + header_name + ":" + s + "\n";
```
For this PR, I'd like to change as little as possible in this code, since it's directly copied from the pre-pipeline shared key auth code in storage_account_client.rs. If there are any issues there, I'd like to address them in a follow-up PR.

thovoll commented:
@rylev I addressed your comments. Most of them I'd like to fix in a follow-up PR since they are in copy/pasted code and I'd like to keep it "original" for this PR and address these in a follow-up PR.

rylev commented on sdk/storage_datalake/src/clients/data_lake_client.rs:
```diff
@@ -1,8 +1,10 @@
-use crate::authorization_policy::AuthorizationPolicy;
+// use crate::bearer_token_authorization_policy::BearerTokenAuthorizationPolicy;
```
Can we remove this? I think the TODO below is sufficient, and we don't need to keep this.

thovoll replied:
Done

rylev reviewed (approved):
Seems fine to me if the nits are addressed in a follow up PR. I'd like for @heaths to review this again though.

thovoll commented on sdk/storage_datalake/src/shared_key_authorization_policy.rs:
```diff
@@ -0,0 +1,244 @@
+        );
+        headers_mut.append(
+            azure_core::headers::VERSION,
+            HeaderValue::from_str("2019-12-12")?,
```
Yes

thovoll commented:
@rylev follow-up PR: https://github.com/Azure/azure-sdk-for-rust/pull/595
